### DISCUSSION TITLE
Options other than GetSemanticModel?

### DISCUSSION THREAD

**Discussion by 7thFox** [dotnet/roslyn#63207] [Q&A]
We're in the process of updating our analyzers to Microsoft.CodeAnalysis 4.2.0 and since our previous version `semanticModel.GetOperation(syntaxNode)` has broken in a place where we attempted to get an Operation from a different SyntaxTree/SemanticModel.

I've managed to make that analyzer rule work again by calling `Compilation.GetSemenaticModel(syntaxTree)`, however I'm wondering if there are any alternative methods to getting an operation outside the current syntax tree (specifically in this case, the body of a method) without [re]creating a new semantic model for that SyntaxTree? Just to head off some options I've already thought of:

1. Syntax analysis - I need type information (and potentially would need to do the same move to a method recursively); both of these would still require a SemanticModel
2. Some use of `RegisterCompilationEndAction` - This has a few issues that are specific to our application, namely the size of solution we're analyzing. Without going in depth, it would take too long to see diagnostics within Visual Studio, and at the end of the day we're still calling GetSemanticModel if we're unlucky and have already completed analysis on that tree.

My current solution is basically just to call GetSemanticModel and give up after a certain depth of traversal. I think this will work ok for us, but I wanted to see if there was any other option I am missing.

Here's some relevant code snippets if it helps give context:

```C#
// What I'm trying to find:

public SomeDelegate DelegateProperty// watching specifically within getters for these delegate properties
{
    get
    {
        new SomeClass();// this class should only get instantiated within the delegate lambda. Error!
        return x => { ... };
    }
}
```

```C#
            context.RegisterOperationBlockStartAction(blockCtx =>
            {
                if (blockCtx.OwningSymbol.Kind == SymbolKind.Method)
                {
                    // Property { get; } returning a SomeDelegate
                    IMethodSymbol methodSymbol = blockCtx.OwningSymbol as IMethodSymbol;
                    if (methodSymbol.MethodKind == MethodKind.PropertyGet && methodSymbol.ReturnType == Constants.SomeDelegate)
                    {
                        blockCtx.RegisterOperationAction(OnObjectCreation, OperationKind.ObjectCreation);// new BusinessField
                    }
                }
            });

...

        private void OnObjectCreation(OperationAnalysisContext context)
        {
            IObjectCreationOperation op = context.Operation as IObjectCreationOperation;
            if (!IsInLambda(op) && CreationInstantiatesSomeClass(op))
            {
                context.ReportDiagnostic(...);
            }
        }

...

        private static bool InstantiatesSomeClass(IOperation op)
        {
                // ...
                case OperationKind.Invocation:
                // ...
                        InvokeInstantiatesSomeClass(op as IInvocationOperation)
                // ...
        }

...

        // Previous version:
        private static bool InvokeInstantiatesSomeClass(IInvocationOperation op)
        {
            if (op.TargetMethod.GetFullyQualifiedName().StartsWith("Some.Namespace."))
            {
                // ...
                return op.TargetMethod.DeclaringSyntaxReferences
                    .Any(sy => InstantiatesSomeClass(
                        op.SemanticModel.GetOperation(sy.GetSyntax()),// Breaks in 4.2.0 -- different syntax tree
                        recheckLambda));
            }

            return false;
        }

        // New version:
        private static bool InvokeInstantiatesSomeClass(IInvocationOperation op, OperationAnalysisContext context)
        {
            if (op.TargetMethod.GetFullyQualifiedName().StartsWith("Some.Namespace."))
            {
                // ...
                return op.TargetMethod.DeclaringSyntaxReferences
                    .Any(sy =>
                    {
                        SemanticModel semanticModel = op.SemanticModel;
                        if (semanticModel.SyntaxTree.FilePath != sy.SyntaxTree.FilePath)
                        {
                            // ...
                            semanticModel = op.SemanticModel.Compilation.GetSemanticModel(sy.SyntaxTree);// sad path
                        }

                        return InstantiatesSomeClass(
                            semanticModel.GetOperation(sy.GetSyntax()),
                            context);
                    });
            }

            return false;
        }
```

**Comment by jasonmalinowski** [comment]
So I'm a bit confused here in your previous version:

```
 return op.TargetMethod.DeclaringSyntaxReferences
                    .Any(sy => InstantiatesSomeClass(
                        op.SemanticModel.GetOperation(sy.GetSyntax()),// Breaks in 4.2.0 -- different syntax tree
                        recheckLambda));
```

I wouldn't expect that code as written to have ever worked reliably. The semantic model is bound to a specific syntax tree, but that code would be assuming that the target method's syntax (which is the declaration syntax) and the operation you're processing is in the same file. My gut is this has nothing to do with 4.2, but you were just getting lucky and you hadn't actually written that case in your project that's consuming this analyzer.

Your new version could be simplified to just something like this (caveat: I have not actually compiled this, so may have typos, etc):

```
        private static bool InvokeInstantiatesSomeClass(IInvocationOperation op, OperationAnalysisContext context)
        {
            if (op.TargetMethod.GetFullyQualifiedName().StartsWith("Some.Namespace."))
            {
                // ...
                return op.TargetMethod.DeclaringSyntaxReferences
                    .Any(sy =>
                    {
                        var semanticModel = context.Compilation.GetSemanticModel(sy.SyntaxTree);

                        return InstantiatesSomeClass(
                            semanticModel.GetOperation(sy.GetSyntax()),
                            context);
                    });
            }

            return false;
        }
```

**Reply by 7thFox** [reply]
>I wouldn't expect that code as written to have ever worked reliably

 Yeah, I don't disagree that this was probably not working as intended before. From what I had read elsewhere it was an intentional change to throw when accessing from another syntax tree instead of doing silent expensive calls. I know it worked previously, however, because I had unit tests specifically to test it's functionality across syntax trees; it was just an intentional break in the API based on my previous research (basically to prevent exactly what I was doing).

>Your new version could be simplified

90% of the time it _is_ on the same syntax tree; always calling `GetSemanticModel` is going to incur a big penalty each time, which is why `RS1030:Do not invoke Compilation.GetSemanticModel() method within a diagnostic analyzer` exists

**Reply by jasonmalinowski** [reply]
> From what I had read elsewhere it was an intentional change to throw when accessing from another syntax tree instead of doing silent expensive calls.

Looking at the code Git blame at least puts the "must be from same tree check" there from when the original operation API was implemented: 

https://github.com/dotnet/roslyn/blame/8c1eb8e3a3e891eae351538d70c4cdf2cce60ea2/src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs#L477

**Reply by 7thFox** [reply]
I took a look back at the unit test I had, it it was testing that it didn't throw, not that it positively identified the diagnostic. I added another test and it did indeed not work; it just silently returned `null` instead of throwing as it now does. Really, regardless of if it worked, my question was more about how I _should_ go about getting this other semantic model.

(if it's of interest, my previous version was Microsoft.CodeAnalysis 4.0.1)

**Comment by CyrusNajmabadi** [comment]
Agree with Jason, this should have never worked (I'm really not sure how it could have). 

Generally speaking, the idea of a semantic modern is that it is:

1. Cheap to acquire. 
2. Lazy to compute data for what is inside of it. 
3. Caches that data to make future requests for the same data within it cheap. 
4. Discarded when no longer necessary. 
5. Can take up lots of memory over time if held onto. 

Because of this, the best approach is to sweep across the Compilation one tree at a time. Get the semantic model for that tree, and do all your processing then.  Then release the tree. 

This prevent excess CPU usage due to redundant computation, and also allows memory to be cleaned up when no longer needed.

**Reply by 7thFox** [reply]
>best approach is to sweep across the Compilation one tree at a time

This is kind of what I was getting at with my comments about `RegisterCompilationEndAction`. Ideally maybe we could do this, but I have no guarantee that I'm going to know I need something from a syntax tree until after I've already gone through it (and thus, I'm still calling `GetSementicModel`, albeit, delayed). Related, you had mentioned caching, but the number of false-positives I'd cache would probably bloat RAM a great deal. Just to give an idea of it, based on my quick search, there's about 10500 instantiations of `SomeClass` in our solution currently.

This did spark the idea of "caching" a list of methods that instantiate  `SomeClass`, but it would run into issues with something like

```C#
public SomeDelegate DelegateProperty// watching specifically within getters for these delegate properties
{
    get
    {
        var y = Method1();// instantiates SomeClass outside lambda. Error!
        return x => { ... };
    }
}

// different syntax tree

public static SomeWrapperClass Method1()
{
    return Method2();
}

public static SomeWrapperClass Method2()
{
    // ...    
    new SomeClass();// this class should only get instantiated within the delegate lambda. Error!
    // ...    
}
```

unless I also cache what methods are called in by method as well.

**Reply by 7thFox** [reply]
If "just call `GetSemanticModel`" is the solution, that's fine too; we really don't have to often currently; `RS1030:Do not invoke Compilation.GetSemanticModel() method within a diagnostic analyzer` just makes me hesitant to.
